From efb0a7e5e78ab261ad8ac5af0ff867fe9ce8c99a Mon Sep 17 00:00:00 2001
From: Michael Niedermayer <michael@niedermayer.cc>
Date: Fri, 28 Aug 2026 16:34:20 +0000
Subject: [PATCH] avformat/mov: clip overly large trun sample durations

The mov demuxer already treats sample durations above max_stts_delta as invalid when reading stts. Apply the same policy to fragmented files before storing trun durations in the packet time-to-sample table.

This prevents a corrupt UINT32_MAX duration from making timestamp-generating filters duplicate millions of frames.

Fixes issue #22843.

Assisted-by: Fairy
(cherry picked from commit efb0a7e5e78ab261ad8ac5af0ff867fe9ce8c99a)

Hades: the FATE test is added after the existing trun test neighbours of
9.0.2's tests/fate/mov.mak; upstream adds it after the entries of
b57a3d028dbe, which 9.0.2 does not have. The change to mov.c is
upstream's.
---
 libavformat/mov.c                             |  7 +++++++
 tests/fate/mov.mak                            | 16 ++++++++++++++++
 tests/ref/fate/mov-trun-large-sample-duration | 10 ++++++++++
 3 files changed, 33 insertions(+)
 create mode 100644 tests/ref/fate/mov-trun-large-sample-duration

diff --git a/libavformat/mov.c b/libavformat/mov.c
index 984c46122ad7..9328245e5d2f 100644
--- a/libavformat/mov.c
+++ b/libavformat/mov.c
@@ -6165,6 +6165,13 @@ static int mov_read_trun(MOVContext *c, AVIOContext *pb, MOVAtom atom)
         if (flags & MOV_TRUN_SAMPLE_FLAGS)    sample_flags    = avio_rb32(pb);
         if (flags & MOV_TRUN_SAMPLE_CTS)      ctts_duration   = avio_rb32(pb);
 
+        if (sample_duration > c->max_stts_delta) {
+            av_log(c->fc, AV_LOG_WARNING,
+                   "Too large sample duration %u in trun entry %u in st:%d. Clipping to 1.\n",
+                   sample_duration, i, st->index);
+            sample_duration = 1;
+        }
+
         mov_update_dts_shift(sc, ctts_duration, c->fc);
         if (pts != AV_NOPTS_VALUE) {
             dts = pts - sc->dts_shift;
diff --git a/tests/fate/mov.mak b/tests/fate/mov.mak
index 56dc3763c003..3450ff0a7e75 100644
--- a/tests/fate/mov.mak
+++ b/tests/fate/mov.mak
@@ -304,6 +304,22 @@ fate-mov-vfr: CMD = md5 -filter_complex testsrc=size=2x2:duration=1,setpts=N*N:s
 fate-mov-vfr: CMP = oneline
 fate-mov-vfr: REF = 1558b4a9398d8635783c93f84eb5a60d
 
+FATE_MOV_FFMPEG_FFPROBE-$(call ALLYES, COLOR_FILTER SETPTS_FILTER MPEG4_ENCODER \
+                                      MP4_MUXER MOV_DEMUXER FILE_PROTOCOL)      \
+                                      += fate-mov-trun-large-sample-duration
+
+# Generate a fragmented VFR file, then replace the last sample duration in
+# its first trun with UINT32_MAX.
+fate-mov-trun-large-sample-duration: CMD = run_with_patched_temp \
+    "$(FFMPEG) -nostdin -v error \
+    -filter_complex color=c=black:s=2x2:r=1,setpts=N*N \
+    -frames:v 10 -fps_mode vfr -c:v mpeg4 -g 5 -bf 0 -q:v 2 -threads 1 \
+    -flags +bitexact -fflags +bitexact \
+    -movflags empty_moov+frag_keyframe+default_base_moof -f mp4 -y" \
+    "ffprobe$(PROGSSUF)$(EXESUF) -show_packets \
+    -show_entries packet=pts,dts,duration -print_format compact \
+    -select_streams v -v 0" mp4 923 "\\0377\\0377\\0377\\0377"
+
 FATE_MOV_FFMPEG_FFPROBE-$(call TRANSCODE, FLAC, MP4 MOV, WAV_DEMUXER PCM_S16LE_DECODER) += fate-mov-mp4-iamf-stereo
 fate-mov-mp4-iamf-stereo: tests/data/asynth-44100-2.wav tests/data/streamgroups/audio_element-stereo tests/data/streamgroups/mix_presentation-stereo
 fate-mov-mp4-iamf-stereo: SRC = $(TARGET_PATH)/tests/data/asynth-44100-2.wav
diff --git a/tests/ref/fate/mov-trun-large-sample-duration b/tests/ref/fate/mov-trun-large-sample-duration
new file mode 100644
index 000000000000..d65fa60996cf
--- /dev/null
+++ b/tests/ref/fate/mov-trun-large-sample-duration
@@ -0,0 +1,10 @@
+packet|pts=0|dts=0|duration=16384
+packet|pts=16384|dts=16384|duration=49152
+packet|pts=65536|dts=65536|duration=81920
+packet|pts=147456|dts=147456|duration=114688
+packet|pts=262144|dts=262144|duration=1
+packet|pts=409600|dts=409600|duration=180224
+packet|pts=589824|dts=589824|duration=212992
+packet|pts=802816|dts=802816|duration=245760
+packet|pts=1048576|dts=1048576|duration=278528
+packet|pts=1327104|dts=1327104|duration=16384
